Add What's New In DevTools (Chrome 80) (#8310)
google/WebFundamentals: src/content/en/updates/2019/12/webassembly.md

### A temporary workaround

To work around this problem, Emscripten and DevTools have temporarily adapted the existing 
[source maps](https://www.html5rocks.com/en/tutorials/developertools/sourcemaps/) format to WebAssembly. This allowed mappings
between binary offsets in the compiled module to original locations in source files.

<figure>
  <img src="../../images/2019/12/sourcemaps.png"
       alt="A screenshot of the source-maps-powered debugging."/>
</figure>

However, source maps were designed for text formats with clear mappings to JavaScript concepts and values, not for binary formats like WebAssembly with arbitrary source languages, type systems, and a linear memory. This made the integration hacky, limited, and not widely supported outside Emscripten.

## Enter DWARF

On the other hand, many native languages already have a common debugging format,
[DWARF](http://dwarfstd.org/), that provides all the necessary information for debuggers to resolve locations, variable names, type layouts, and more.

> While there are still some WebAssembly-specific features that need to be added for full compatibility, compilers like Clang and Rust already support emitting DWARF information in WebAssembly modules, which enabled the DevTools team to start integrating DWARF information

Reviewer: Hmm I'm not sure "integrating DWARF information" makes sense as it's just information and is part of user-supplied Wasm modules, not DevTools. I'd leave this either as in original draft

```suggestion
While there are still some WebAssembly-specific features that need to be added for full compatibility, compilers like Clang and Rust already support emitting DWARF information in WebAssembly modules, which enabled the DevTools team to start using it
```

or maybe say

```suggestion
While there are still some WebAssembly-specific features that need to be added for full compatibility, compilers like Clang and Rust already support emitting DWARF information in WebAssembly modules, which enabled the DevTools team to start integrating DWARF support
```

or something similar.

directly into DevTools.

> As a first step, the Chrome DevTools team implemented native source mapping using this information, so you can start debugging Wasm modules produced by any of these compilers without resorting to the disassembled format or having to use any custom scripts.

Reviewer: I don't know if anyone cares about these details... but I'm not on the Chrome DevTools team, so I just carefully said "we" (as a wider team) in the draft 😅 Maybe worth keeping that way?

```suggestion
As a first step, we implemented native source mapping using this information, so you can start debugging Wasm modules produced by any of these compilers without resorting to the disassembled format or having to use any custom scripts.
```

> Instead, you just need to tell your compiler to include debug info like you normally would on other platforms. For example, in Clang this can be done by passing the `-g` flag during compilation:

Reviewer: I'd add a newline / start a new paragraph here:

```suggestion

Instead, you just need to tell your compiler to include debug info like you normally would on other platforms. For example, in Clang this can be done by passing the `-g` flag during compilation:
```

    clang -g ...sources… -target wasm32 -o out.wasm

You can use same `-g` flag in Rust:

    rustc -g source.rs --target wasm32-unknown-unknown -o out.wasm

> If you’re using Cargo, the debug info will be included by default:

Reviewer: Cargo is also Rust (just a more general builder), so this sentence is meant to be a continuation of the last one. I probably haven't made that very clear in the draft, so maybe like this it will work better:

```suggestion
...or, if you’re using Cargo, the debug info will be included by default:
```

    cargo build --target wasm32-unknown-unknown

This new DevTools integration with DWARF already covers support for stepping over the code, setting breakpoints, and resolving stack traces in your source languages.

<figure>
  <img src="../../images/2019/12/dwarf.png"
       alt="A screenshot of the new DWARF-powered debugging."/>
</figure>

## The future

There is still quite a bit of work to do though.
For example, on the tooling side, Emscripten (Binaryen) and wasm-pack (wasm-bindgen) don’t support updating DWARF information on transformations they perform yet. For now, they won’t benefit from this integration. 

And on the Chrome DevTools side, we’ll be evolving integration more over time to ensure a seamless debugging experience, including: 

* Resolving variable names
* Pretty-printing types

> * Evaluating expressions in source languages

Reviewer: ```suggestion
* Evaluating expressions in source languages
* ...and much more!
```

(as these are just examples)
